beacon_node/network/src/sync/block_lookups/mod.rs

```diff
@@ -419,269 +410,229 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
-                                    cx,
-                                ) {
-                                    Ok(()) => ShouldRemoveLookup::False,
-                                    Err(()) => ShouldRemoveLookup::True,
-                                }
-                            })
-                            .unwrap_or(ShouldRemoveLookup::False)
-                    } else {
-                        ShouldRemoveLookup::False
-                    }
-                } else {
-                    // These are the correct blobs, send them for processing
-                    match self.send_blobs_for_processing(
+    fn handle_verified_response<L: Lookup, R: RequestState<L, T>>(
+        &self,
+        seen_timestamp: Duration,
+        cx: &SyncNetworkContext<T>,
+        process_type: BlockProcessType,
+        verified_response: R::VerifiedResponseType,
+        lookup: &mut SingleBlockLookup<L, T>,
+    ) -> Result<(), LookupRequestError> {
+        let id = lookup.id;
+        let block_root = lookup.block_root();
+
+        R::request_state_mut(lookup)
+            .get_state_mut()
+            .component_downloaded = true;
+
+        let cached_child = lookup.add_response::<R>(verified_response.clone());
+        match cached_child {
+            CachedChild::Ok(block) => {
+                // If we have an outstanding parent request for this block, delay sending the response until
+                // all parent blocks have been processed, otherwise we will fail validation with an
+                // `UnknownParent`.
+                let delay_send = match L::lookup_type() {
+                    LookupType::Parent => false,
+                    LookupType::Current => self.has_pending_parent_request(lookup.block_root()),
+                };
+
+                if !delay_send {
+                    self.send_block_for_processing(
                         block_root,
-                        blobs,
+                        block,
                         seen_timestamp,
-                        BlockProcessType::SingleBlock { id },
+                        process_type,
                         cx,
-                    ) {
-                        Ok(()) => ShouldRemoveLookup::False,
-                        Err(()) => ShouldRemoveLookup::True,
-                    }
+                    )?
                 }
             }
-            Ok(None) => ShouldRemoveLookup::False,
-            Err(e) => handle_block_lookup_verify_error(
-                request_ref,
-                ResponseType::Blob,
-                peer_id,
-                e,
+            CachedChild::DownloadIncomplete => {
+                // If this was the result of a block request, we can't determined if the block peer
+                // did anything wrong. If we already had both a block and blobs response processed,
+                // we should penalize the blobs peer because they did not provide all blobs on the
+                // initial request.
```

Comment:
Are peers required to provide all blobs? what if they only have a subset of blobs in a block, or if they limit the number of blobs returned in a single response? can we request from other peers for the remaining blobs if we only get a subset of blobs?

> Clients MAY limit the number of blocks and sidecars in the response.

https://github.com/ethereum/consensus-specs/blob/36d9ea6cd19963ddf0f3628b438e85dc6acddf64/specs/deneb/p2p-interface.md?plain=1#L292

Reply:
Thanks for the explanation!